src/main.c

```diff
@@ -5,380 +5,80 @@
-	}
-
-	ret = strlen(root_obj_string);
-	memcpy(buffer, root_obj_string, ret);
-	cJSON_FreeString(root_obj_string);
-
-exit:
-	cJSON_Delete(root_obj);
-
-	return ret;
-}
+volatile enum lte_lc_nw_reg_status network_status;
 
-static int lte_connection_check(void)
+int get_network_status(void)
 {
-	int err;
-
-	enum lte_lc_nw_reg_status nw_reg_status;
-
-	err = lte_lc_nw_reg_status_get(&nw_reg_status);
-	if (err) {
-		printk("lte_lc_nw_reg_status_get, error: %d\n", err);
-		return err;
-	}
-
-	switch (nw_reg_status) {
-	case LTE_LC_NW_REG_REGISTERED_HOME:
-	case LTE_LC_NW_REG_REGISTERED_ROAMING:
-		return 0;
-	default:
-		return -1;
-	}
+	return network_status;
 }
 
-int send_data(int client_fd, int timeout_s)
+static void lte_handler(const struct lte_lc_evt *const evt)
 {
-	int err;
-	char send_buf[SEND_BUF_SIZE] = { 0 };
-	int send_len;
-	struct modem_param_info modem_params = { 0 };
-
-	err = modem_info_params_init(&modem_params);
-	if (err) {
-		printk("Modem info params could not be initialised: %d\n", err);
-		return -1;
-	}
-
-	err = modem_info_params_get(&modem_params);
-	if (err < 0) {
-		printk("Unable to obtain modem parameters: %d\n", err);
-		return -1;
-	}
-
-	send_len = create_send_buffer(&modem_params, send_buf, timeout_s);
-	if (send_len < 0) {
-		printk("Error creating json object: %d\n", send_len);
-		return -1;
-	}
-
-	err = send(client_fd, send_buf, send_len + 1, 0);
-	if (err < 0) {
-		printk("Failed to send data, errno: %d\n", errno);
-
-		return 0;
-	}
-	printk("Packet sent: %s\n", send_buf);
-	return 1;
-}
-
-int poll_and_read(int client_fd, int timeout_s)
-{
-	int err;
-	char recv_buf[RECV_BUF_SIZE] = { 0 };
-	size_t ret_len;
-	struct pollfd fds[] = { { .fd = client_fd, .events = POLLIN } };
-	s64_t start_time = k_uptime_get();
-	s64_t total_poll_time_ms = 0;
-
-	while (1) {
-		err = poll(fds, ARRAY_SIZE(fds), POLL_TIMEOUT);
-		if (err < 0) {
-			printk("poll, error: %d", err);
-
-			return -1;
-		} else if (err == 0) {
-			total_poll_time_ms += k_uptime_delta(&start_time);
-			if (total_poll_time_ms <
-			    (timeout_s + TIMEOUT_TOL) * S_TO_MS_MULT) {
-				printk("Elapsed time: %lld seconds\n",
-				       total_poll_time_ms / S_TO_MS_MULT);
-				continue;
-			}
-
-			printk("No response from server\n");
-			return 0;
-		} else if ((fds[0].revents & POLLIN) == POLLIN) {
-			ret_len = recv(client_fd, recv_buf,
-				       sizeof(recv_buf) - 1, 0);
-			if (ret_len > 0 && ret_len < SEND_BUF_SIZE) {
-				// Received "invalid format" message
-				if (!strcmp(recv_buf, INV_FORMAT_MSG)) {
-					printk("Response: %s\n", recv_buf);
-					return -1;
+	static int reconnect_counter;
+
+	switch (evt->type) {
+	case LTE_LC_EVT_NW_REG_STATUS:
+		switch (evt->nw_reg_status) {
+		case LTE_LC_NW_REG_REGISTERED_HOME:
+		case LTE_LC_NW_REG_REGISTERED_ROAMING:
+			reconnect_counter = 0;
+			break;
+		case LTE_LC_NW_REG_SEARCHING:
+			break;
+		default:
+			if (reconnect_counter <
+			    CONFIG_NAT_TEST_MAX_LTE_RECONNECT_ATTEMPTS) {
```

Comment:
I'm not sure about this condition. The `reconnect_counter` will be incremented for every `LTE_LC_EVT_NW_REG_STATUS` event that doesn't match any of the cases above. But it doesn't really count connection attempts, it counts the number of events ending up in the `default` case.

Reply:
The idea was: "If disconnected and not exhausted reconnect attempts -> attempt to reconnect"
I count all cases other than registered and searching as a disconnect/error which then can trigger a reconnect attempt.
Let me know if my thinking is wrong here though. I will have to

Reply:
It's possible to have multiple `LTE_LC_NW_REG_SEARCHING` and `LTE_LC_NW_REG_UNKNOWN` before finding a suitable cell to connect to. I think I'd check for cases where service is denied or search fails, otherwise let the search continue until timeout. After that I think a call to `lte_lc_normal` would not do much, as the radio is already activated.

Reply:
Refer to my earlier comment regarding the change of this event handler